### ISSUE TITLE
K8SSAND-615 ⁃ Add support for adding a datacenter to an existing cluster

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/k8ssandra-operator#21]
## Background
The steps for adding a DC to an existing cluster differ from those for setting up a new, multi-DC cluster. It involves updating replication strategies of keyspaces and streaming data from an existing DC to the new DC. See the following for more background on the steps involved for adding a DC to an existing cluster:

* https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/operations/opsAddDCToCluster.html
* https://thelastpickle.com/blog/2019/02/26/data-center-switch.html

## Purpose
The purpose of this ticket is to provide support for adding a DC to an existing cluster where the DC will live in a different k8s cluster. 

We will also support adding a DC within an existing k8s cluster, but #95 needs to be done first. Adding a DC within an existing k8s cluster will therefore be addressed in a separate ticket.

## Details
Assume _DC_  refers to a CassandraDatacenter unless noted otherwise.

* Need to distinguish between adding a DC to a cluster that is initializing vs one that is fully initialized
  * If a DC is added to the K8ssandraCluster before all DCs are initialized then the process is the same as setting up a new, multi-DC cluster
  * If a DC is added to the K8ssandraCluster after all DCs are initialized then we follow the process for adding a DC to an existing cluster
  * We can determine if the cluster is initialized by
    * Add a `CassandraInitialized` status condition to K8ssandraCluster
    * Set `CassandraInitialized` to `true` after all DCs have their `Initialized condition set to `true` 
* All user-created keyspaces as well as system_traces, system_distributed, and system_auth should be configured to use NetworkTopologyStrategy before adding a new DC
  * Should we add a validation check for user-created keyspaces in an effort to prevent misconfigured clusters?
* Deploy the new DC
  * Configure additional seeds as we normally do so that the new DC has seeds from other, existing DCs
* Wait for the DC to reach the Ready state
* Update replication strategy for all user-created keyspaces as well as system_traced, system_distributed, and system_auth to include replicas for the new DC
  * Need to provide a way for user to specify replicas for each keyspace and each DC
* Run rebuild on each node in new DC
  * Provide a way for user to specify src DC for streaming
  * Rebuild can run on more than one node at a time. This will be left as an optimization.

#### Should we add a validation check for user-created keyspaces in an effort to prevent misconfigured clusters?
I don't think there is a sound reason to use SimpleStrategy in a multi-DC cluster. k8ssandra-operator doesn't have a way to completely prevent it; therefore, I do not think a validation check should be added as part of the normal reconciliation process. I do however think that when adding a new DC the operator should check and only proceed if all user keyspaces are using NTS.

#### How can the user specify replicas for each keyspace and each DC?
This is tricky because k8ssandra-operator isn't managing the replication strategies of keyspaces, but it does need to update the replication as part of the workflow for adding a new DC. It is also complicated by the fact that in Cassandra 4 you cannot specify DCs that do not exist. You can do this in Cassandra 3. This would simplify things because the user could simply update the replication prior to adding the new DC to the K8ssandraCluster.

I am not in favor of specifying the info in the K8ssandraCluster spec because it involves details that k8ssandra-operator doesn't really manage. I think this could create some confusion.

Another option is for the user to specify the replication via a custom type, e.g., `CassandraReplicationStrategy` or ConfigMap. Here is an example of how a ConfigMap might look:

```yaml
apiVersion: v1
kind: ConfigMap
data:
  replication: |-
    - keyspace: ks_1
      dataceteners:
        "dc1": 3
        "dc2": 3
        "dc3": 3
    - keyspace: ks_2
      dataceteners:
        "dc1": 3
        "dc2": 3
        "dc3": 3
```

The operator could use this ConfigMap if it exists. If it doesn't exist, then the operator could apply a default like `replication_factor = min(dc_size, 3)`.

#### How will this impact existing orchestration workflows?
We only reconcile Stargate (and Reaper as well with #216) after reconciling all CassandraDatacenters. For example, if we have 3 DCs - dc1, dc2, and dc3 - K8ssandraClusterReconciler will reconcile Stargate if each of the DCs is a ready state. Now let's say we add dc4 and in dc2 we want to add or remove Stargate. Currently the Stargate changes won't be made in dc2 until dc4 is ready. This is suboptimal at best and needs to be changed.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-615) by [Unito](https://www.unito.io)
┆Reviewer: 5fa4593062584c006b1e04d0
┆epic: Addition of new DataCenters
┆fixVersions: k8ssandra-operator-v1.0.0
┆friendlyId: K8SSAND-615
┆priority: Medium

**Comment by jsanda** [comment]
management-api PR to add endpoint for rebuild operations https://github.com/k8ssandra/management-api-for-apache-cassandra/pull/161

**Comment by jsanda** [comment]
I have a concern about specifying replication for all keyspaces in the ConfigMap approach. Suppose the user inadvertently specifies the wrong number of replicas for another, existing DC. The operator then proceeds to change the replica count for that DC. That would be bad. We could avoid this situation by having the operator ignore the other DCs, but then it seems pointless and confusing to have the user specify them. 

Given all this I think a better and simpler approach would be to specify the desired replication for the new DC via an annotation on the K8ssandraCluster object. It could look something like this:

```yaml
apiVersion: k8ssandra.io/v1alpha1
kind: K8ssandraCluster
metadata:
  annotations:
    k8ssandra.io/dc-replication: '{"dc2": {"ks1": 3, "ks2": 3}}'
```
The value is a map serialized as JSON. This would allow multiple DCs to be specified which would address scenarios where multiple DCs are added to the K8ssandraCluster in one update. 

The behavior I am proposing is as follows:

* The operator only uses this annotation when a new DC is added to an existing cluster. It is otherwise ignored.
* If specified the operator applies replication changes only for the specified keyspaces
* The operator applies replication changes only if the DC is not already part of the cluster
    * Cluster membership in this context will be determined by looking at the K8ssandraCluster status. If the DC is reported in the status, then no replication updates should be applied.
* If the annotation is not specified or if the new DC is not in the annotation value, then the operator should compute the replication for all user-defined keyspaces as follows, `replication_factor = min(size(new_dc), 3)`

**Comment by adejanovski** [comment]
I like the idea of using the annotation.
I'm a little more nervous about having default behaviors for keyspaces that aren't explicitly specified in the annotation.
One could assume (and that would be legit) that nothing happens to keyspaces that aren't in the annotation, although they'll be modified to replicate on the new DC.

Maybe we could start with more safety over convenience, and expect all existing keyspaces replication factors to be set in the annotation (0 meaning unreplicated) and fail the reconcile if there are still non system keyspaces which aren't there.

We can revisit that and define default behaviors based on user feedback.

wdyt?

**Comment by jsanda** [comment]
This will help eliminate surprises. Sounds good to me.

**Comment by burmanm** [comment]
So hopefully I remember everything we discussed, but here's my take - I want to see more automation rather than less. Not all K8ssandra users are supposed to be experts and also managing a JSON in annotation is painful and slow.

What I'd rather see is a set of options (even in a custom type), such as:

* automate
  * k8ssandra-operator sets all the replication_factors based on our logic
* user-defined: |-
  * user has to set all the user keyspaces
* overrides: |-
  * Take the input from automate or default if the keyspace isn't in this override list
* default: n
  * unless overridden, the default value set here is used

With the truth table looking something like this in a table (x = error, * = accept):

|    | automatic | user-defined | override | default |
|----|---|----|----|---|
| automatic | - | x  | *  | x |
| user-defined | x | -  | x  | x |
| override | * | x  | -  | * |
| default | x | x  | *  | - |

Thus more than 2 set is never accepted.

**Comment by jdonenine** [comment]
So catching up on this, I had a few thoughts I wanted to get down...

A really good goal for us to trek towards as a project is to automate as much as possible and make C* itself easier to manage in the process -- it's naive to think we can remove all of the knowledge necessary to operate it, but we can take steps like what @burmanm proposes there to make it easier -- I love the idea of the `automate` setting, where you can in theory set it and forget it.

That I think though is not where we're at right now in our initial phase, so what I'd propose we do is take the page from what is coined there as `user-defined` and say that's where we start.  We'll expand later to include other optional approaches (I'll create an additional ticket to track that idea so we don't lose it), but we need to start with an approach we can manage timeline wise.

On the topic of annotation vs structured resource though, I'm quite torn.  One one hand I can see values (like self-documentation and potential validation) of a structured resource -- but I also dislike the increased proliferation of configuration as more and more resources are created.  As a user in this case I feel like I'd like to set that configuration directly in the K8ssandraCluster resource -- in fact I'd like to set as much there as possible really.  The annotation approach on the other hand gives that locality of configuration but lacks in the other ways and is probably a bit more error prone in nature -- although I'm not too terribly averse to the JSON string myself, but maybe that's just because I'm used to it in the web application world (hey, at least it's not yaml - am I right? 😉 ).

My main concern here would be are we creating anything that we can't come back from -- knowing that our end-goal is likely to be something closer to what is described with the `automate` workflow.  It would seem to me that the annotation based approach could be phased out later and transitioned over to a structured resource if necessary -- it would mean some additional logic within the controller, but not any additional burden on the user, and that I think we can live with.  

I'd move in favor of the more simple approach we currently have and get that over the finish line.  Once that is getting some usage we can think a bit more about how we can enhance it, and instead of focusing on the input mechanism (structured vs annotation) instead dedicate time to solving the automation case -- which I think has a lot more benefit to users in the long run.
